### PR TITLE
Fix: enable non .txt file format uploads when using direct upload

### PR DESCRIPTION
**Pull request by rjanjua**
### Related Issues

- fixes an issue raised by Sol on behalf of Cambridge - pdf (and other non text formats) uploads do not work for direct uploads (https://deepset.atlassian.net/jira/polaris/projects/DCD/ideas/view/3949566?selectedIssue=DCD-184&issueViewSection=comments)

### Proposed Changes?

Add a new function in the sync and async clients - `upload_bytes`, that allows the upload of any supported formats. Updated some internal implementations of the sdk to enable this without replicating the same code for different formats.

 

### Checklist

- [ ] I have updated the referenced issue with new insights and changes
- [ ] If this is a code change, I have added unit tests
- [ ] I've used the [conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title
- [ ] I updated the docstrings
- [ ] If this is a code change, I added meaningful logs and prepared Datadog visualizations and alerts

### CODE REVIEW

File: deepset_cloud_sdk/workflows/sync_client/files.py [hunk 1]
```diff
@@ -167,6 +171,45 @@ def upload_texts(
...
+    files: List[DeepsetCloudFileBytes],
+    api_key: Optional[str] = None,
+    api_url: Optional[str] = None,
+    workspace_name: str = DEFAULT_WORKSPACE_NAME,
+    write_mode: WriteMode = WriteMode.KEEP,
+    blocking: bool = True,
+    timeout_s: Optional[int] = None,
+    show_progress: bool = True,
+) -> S3UploadSummary:
+    """Upload texts to deepset Cloud.
```
Comment (sjrl) [inline deepset_cloud_sdk/workflows/sync_client/files.py:184]: I'd update the doc string here to reflect that all file types are supported. It'd be great to list the supported file types here.

File: deepset_cloud_sdk/workflows/sync_client/files.py [hunk 2]
```diff
@@ -167,6 +171,45 @@ def upload_texts(
...
+
+    :param files: List of DeepsetCloudFiles to upload.
```
Comment (sjrl) [inline deepset_cloud_sdk/workflows/sync_client/files.py:186]: Change to `DeepsetCloudFileBytes`

**Comment by sjrl** [comment]
Thanks for this @rjanjua! One clarification question I wanted to ask is what does the `lt 20` mean in the title?

**Comment by sjrl** [comment]
> Ah sorry, updated the title to make more sense - lt = less than

Thanks! Does this mean that the upload of DeepsetCloudFile and DeepsetCloudFileBytes only works when you are uploading less than 20 at a time?

**Comment by rjanjua** [comment]
> > Ah sorry, updated the title to make more sense - lt = less than
> 
> Thanks! Does this mean that the upload of DeepsetCloudFile and DeepsetCloudFileBytes only works when you are uploading less than 20 at a time?

uploading non text files already works batches larger than 20 files. When doing less than 20 files we use direct uploads via the API, but for larger batches we use the upload sessions. This was implemented in a [previous PR](https://github.com/deepset-ai/deepset-cloud-sdk/pull/171), but it seems like the case of less than 20 files was missed somehow.

**Comment by sjrl** [comment]
Thanks for the information. 

To follow-up on this point. 
> uploading non text files already works for batches larger than 20 files.

I'm still a little confused. Could you provide a code snippet of how when using a batch size larger than 20 files that we could directly upload using DeepsetCloudFiles? Or are you saying the DeepsetCloudFiles is only used when doing direct uploads? 

I guess I'm coming from the perspective of using the the python function directly and not the CLI, so from what I can tell your function `upload_bytes` would allow me to upload an arbitrary number of DeepsetCloudFileBytes, which wasn't possible before right?

**Comment by rjanjua** [comment]
I created the class DeepsetCloudFileBytes only because the DeepsetCloudFile class has a field called 'text'. Technically you can populate that field with a bytes object if you so wish, but the naming is confusing. This is why I created the new class with a bytes field instead of a text field. This makes it clear to the user which to use where, even though the underlying way both are handled is essentially the same.

Previously the reason you couldn't upload a single pdf file, for example, was because the direct upload was explicitly checking that files were text files before uploading them, now we allow any file with an extension that is allowed in dC to be uploaded. 

However, when uploading batches of files greater than 20 files, we were uploading via upload sessions, which was checking that the extension was one of the allowed dC extensions. For this reason you could have use upload_texts to upload a pdf file, but only if it was in a batch of more than 20 files.

To reiterate, I created the DeepsetCloudFileBytes class so that it makes more sense to a user when uploading non text files.

**Comment by sjrl** [comment]
Ahh okay, thanks a lot for the in depth explanation @rjanjua!

**Comment by rjanjua** [comment]
> Ahh okay, thanks a lot for the in depth explanation @rjanjua!

If you want to see a demonstration of what is happening in the problem version of the API check out [this colab](https://colab.research.google.com/drive/1RpC1jlkYpztd5-xc-WyIOotwmnqu0HVw#scrollTo=MN_Q2LOM0blb), if you add remove 1 element from the list of files you will see that it does not work, but with 21 files it does.

File: deepset_cloud_sdk/_s3/upload.py [hunk 3]
```diff
@@ -283,13 +283,13 @@ async def upload_texts(
...
                 file_name = file.name
-                tasks.append(self.upload_from_string(file_name, upload_session, file.text, client_session))
+                tasks.append(self.upload_from_memory(file_name, upload_session, file.content(), client_session))
 
                 # meta
                 if file.meta is not None:
                     meta_name = f"{file_name}.meta.json"
-                    metadata = json.dumps(file.meta)
-                    tasks.append(self.upload_from_string(meta_name, upload_session, metadata, client_session))
+                    metadata = bytes(json.dumps(file.meta), encoding="utf-8")
```
Comment (ArzelaAscoIi) [inline deepset_cloud_sdk/_s3/upload.py:291]: why does the metadata need to be bytes too ? can't we just keep the dict here and convert it at the last step ?

File: pyproject.toml [hunk 4]
```diff
@@ -51,6 +51,7 @@ path = "deepset_cloud_sdk/__about__.py"
 tests-with-cov = "pytest --cov-report=term-missing --cov-config=pyproject.toml --cov=deepset_cloud_sdk tests/unit"
 tests-unit = "pytest --cov-report=term-missing --cov-config=pyproject.toml --cov=deepset_cloud_sdk tests/unit"
 tests-integration = "pytest tests/integration"
+try = "python try.py"
```
Comment (ArzelaAscoIi) [inline pyproject.toml:54]: do we need this ?

File: try.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import httpx
+
+from deepset_cloud_sdk.workflows.sync_client.files import (
+    DeepsetCloudFileBytes,
+    WriteMode,
+    upload,
+    upload_bytes,
+)
+
+api_key = "api_eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIwYjYwYzE2YS00NjJlLTRkNmMtYjg2ZS1mYTI3M2RhNGU2Nzh8NjI5NzY4NWVmMzdiNTIwMDY4NDU3ODMxIiwiZXhwIjoxNzE2NTA1MjAwLCJhdWQiOlsiaHR0cHM6Ly9hcGkuZGV2LmNsb3VkLmRwc3QuZGV2Il19.MjfWK8JtrkVpSLwb-wJ06vxkw2sQM3sB6Okd47PX3Cw"
```
Comment (ArzelaAscoIi) [inline try.py:10]: 🔔 🔔 credentials detected => let's drop and invalidate the token

**Review by ArzelaAscoIi — COMMENTED** [review]
Looks good! Could you add a very simple integration test that upload a binary file from memory with metadata and checks that this file exists ? :)  
Otherwise great

File: tests/integration/service/test_integration_files_service.py [hunk 6]
```diff
@@ -212,7 +212,7 @@ async def test_upload_texts(self, integration_config: CommonConfig, workspace_na
                 DeepsetCloudFile("file4", "file4.txt", {"which": 4}),
                 DeepsetCloudFile("file5", "file5.txt", {"which": 5}),
             ]
-            result = await file_service.upload_texts(
+            result = await file_service.upload_in_memory(
```
Comment (ArzelaAscoIi) [inline tests/integration/service/test_integration_files_service.py:215]: so basically here an other test for the other memory file class

File: deepset_cloud_sdk/_s3/upload.py (hunk 3 above)
Comment (rjanjua) [reply deepset_cloud_sdk/_s3/upload.py:291]: even previously we were converting to string, but the upload from memory step only takes bytes. Will update for it to accept string too.

File: pyproject.toml (hunk 4 above)
Comment (rjanjua) [reply pyproject.toml:54]: removed

File: try.py (hunk 5 above)
Comment (rjanjua) [reply try.py:10]: oops!

File: tests/integration/service/test_integration_files_service.py (hunk 6 above)
Comment (rjanjua) [reply tests/integration/service/test_integration_files_service.py:219]: added a pdf file in the above list of files using the DeepsetCloudFileBytes class

**Review by FHardow — APPROVED** [review]
Takeing over from @ArzelaAscoIi.
Skimmed through it and LGTM!

**Pull request merged**